relayer/relayer/src/listeners/azero.rs

```diff
@@ -1,238 +1,158 @@
-                            Err(err) => {
-                                error!("Error in event handler task: {}", err);
-
-                                // Wait for currently active block tasks to finish before restarting the listener
-                                while let Some(result) = block_sealing_tasks.join_next().await {
-                                    _ = result?;
-                                }
-
-                                match err {
-                                    AzeroListenerError::EthContractReverted => {
-                                        if most_eth.paused().call().await? {
-                                            warn!("Most contract on Ethereum is halted, pausing event handling");
-
-                                            wait_until_unpaused(&most_eth).await?;
-
-                                            // Simplest way to recover from a halted bridge is to restart the azero listener
-                                            return Err(
-                                                AzeroListenerError::BridgeHaltedRestartRequired,
-                                            );
-                                        } else {
-                                            return Err(err);
-                                        }
-                                    }
-                                    _ => return Err(err),
-                                }
-                            }
+                    let filtered_events = all_events
+                        .into_iter()
+                        .flat_map(|(block_details, events)| {
+                            most_azero.filter_events(events, block_details.clone())
+                        })
+                        .collect::<Vec<ContractEvent>>();
+
+                    if !filtered_events.is_empty () {
+                        let (events_ack_sender, events_ack_receiver) = oneshot::channel::<()>();
+
+                        select! {
+                            cb_event = circuit_breaker_receiver.recv () => {
+                                warn!(target: "AlephZeroListener", "Exiting before sending events due to a circuit breaker event {cb_event:?}");
+                                return Ok(cb_event?);
+                            },
+
+                            Ok (_) = azero_events_sender
+                                .send(AzeroMostEvents {
+                                    events: filtered_events.clone (),
+                                    events_ack_sender,
+                                }) => {
+                                    info!(target: "AlephZeroListener", "Sending a batch of {} events", &filtered_events.len());
+                                },
                         }
+
+                        info!(target: "AlephZeroListener", "Awaiting events ack");
+                        events_ack_receiver.await.map_err(|_| AlephZeroListenerError::AckSenderDropped)?;
```

Comment:
Question: this blocks until all transactions from that batch are finalized, right?
That would mean that bridge does not listen on Aleph Zero and submit eth transcactions until the whole previous batch is finalized. This isn't ideal, as it doubles (or more) bridging duration of many transactions

Reply:
indeed, when re-writing this part I forgot it waits for a whole era, this indeed is a problem. I'll think of a refactor that always proceeds but only moves the cache forward.

Reply:
But it processes batches in parallel, so most of the time, worst-case would be that transfer takes 2x more time, so it's acceptable + it would be a lot of work to rewrite it.
I think we can keep it that way

Reply:
I think it's not that hard actually - we could:
- spawn independent tasks for batches of events
- use another channel to communicate when to seal a block
- channels are FIFO so the order will be preserved

Reply:
If you see an easy way to do it, then I think it's worth to change it

Reply:
done